### ISSUE TITLE
Autograd deep copy avoidance optimization unsound in the presence of views

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#34082]
Right now, we have the following optimization in accumulate grad:

```
    // under following condition, we can avoid clone()
    if (!GradMode::is_enabled()
        && !new_grad.is_sparse()
        && new_grad.is_contiguous()
        && new_grad.use_count() <= 1 + !post_hooks().empty()) {
      // first check it is in first-order grad only mode
      // then check not sparse before is_contiguous
      // then check contiguous, otherwise later in place accumulation may fail
      // and lastly, check it is the last reference before we grab it.
      // If the function has post hooks (for example, a DDP allreduce hook),
      // call_function in Engine.cpp will temporarily bump the refcount by one, hence the
      // addition of !post_hooks().empty().
      variable.grad() = new_grad.detach();
```

We test the use count on the Tensor object to see if we can steal it into the input buffer (to be accumulated into when we make subsequent modifications). However, this is unsound if `new_grad` is a view on another tensor, because then we will save a view into `variable.grad()` and the aliasing is now observable.

There is also a minor correctness problem, where not all post hooks are safe to apply; if the post hook saves the gradient directly (without copying it) it can also observe subsequent mutations.

cc @mcarilli

**Comment by albanD** [comment]
### The view problem
I agree that this is dangerous.
I kind of remember discussion of moving this `use_count()` check to the storage. But I can't find it anymore (maybe it was just in my head a long time ago :D ).

### The hook problem
The hooks should **never** change the grad inplace and should `.clone()` them if they want to save them.
So in that sense, the hook is not an issue.
Hopefully everyone's hook is actually doing that !

**Comment by ezyang** [comment]
But you can't just `use_count` on storage, because `storage.use_count() == 1` if you have a single tensor referring to the storage, which itself has nonzero use count.

I don't mind saying that the hook problem is not actually a problem.

**Comment by albanD** [comment]
I guess you want both then:
- The backing storage should have a single reference
- The Tensor that uses this storage should have a single use count as well

One thing that can be a bit dangerous with that it does not take into account the weak pointers that could exist.
Do we ever create weak refs to Tensor though?

**Comment by ezyang** [comment]
I think very rarely (mcarilli's patch is doing it). For weak pointers I don't think I care about handling them correctly.
